presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -561,6 +545,75 @@ private void invokeNoMoreSplitsIfNecessary()
         return COMPLETED_FUTURE;
     }
 
+    private Iterator<InternalHiveSplit> buildManifestFileIterator(Table table,
+            FileStatus symlinkFile,
+            FileSystem fileSystem,
+            DirectoryLister directoryLister,
+            NamenodeStats namenodeStats,
+            InternalHiveSplitFactory splitFactory,
+            boolean splittable)
+    {
+        // Strategy: List once, filter file listing based on manifest file. Complication is that not all files in the manifest are guaranteed
+        // to have a common root.
+
+        List<Iterator<LocatedFileStatus>> fileStatusIterators = new ArrayList<>();
+        List<Path> paths = new ArrayList<>();
+
+        try (BufferedReader reader = new BufferedReader(new InputStreamReader(fileSystem.open(symlinkFile.getPath()), StandardCharsets.UTF_8))) {
+            CharStreams.readLines(reader).stream()
+                    .map(Path::new)
+                    .forEach(paths::add);
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_BAD_DATA, "Error parsing symlinks from: " + symlinkFile, e);
+        }
+
+        ImmutableSet<Path> parents = paths.stream()
+                .map(Path::getParent)
+                .distinct()
+                .collect(toImmutableSet());
+
+        for (Path parent : parents) {
+            try {
+                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, parent);
+                fileStatusIterators.add(new HiveFileIterator(table, parent, targetFilesystem, directoryLister, namenodeStats, IGNORED, true));
+            }
+            //There will be a single listing performed if all paths have a common parent. Worst case behavior is N listings for N files
```

Comment:
Could this actually reduce performance. E.g listing `*` could take more time than listing single file.
How often there are different parents?

Reply:
In ad hoc testing, the number of objects returned does not appear to have a significant impact on the response time. There are some cases where it would be worse, for example if the number of objects under a prefix exceeded the page size for an s3 `listObjectsV2` request (1000 objects). 

IME, the common case is for there to be a single parent. This is true for manifest files generated for the partitions of a Delta table.